packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -0,0 +1,129 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { EventEmitter } from "events";
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { assert } from "@fluidframework/common-utils";
+import { IConnectionDetails } from "@fluidframework/container-definitions";
+import { ProtocolOpHandler, Quorum } from "@fluidframework/protocol-base";
+import { ConnectionMode } from "@fluidframework/protocol-definitions";
+import { connectEventName, ConnectionState, ILocalSequencedClient } from "./container";
+
+export class ConnectionStateHandler {
+    private _connectionState = ConnectionState.Disconnected;
+    private _pendingClientId: string | undefined;
+    private _clientId: string | undefined;
+
+    public get connectionState(): ConnectionState {
+        return this._connectionState;
+    }
+
+    public get connected(): boolean {
+        return this.connectionState === ConnectionState.Connected;
+    }
+
+    public get clientId(): string | undefined {
+        return this._clientId;
+    }
+
+    public get pendingClientId(): string | undefined {
+        return this._pendingClientId;
+    }
+
+    constructor(
+        private readonly protocolHandler: () => ProtocolOpHandler | undefined,
+        private readonly logConnectionStateChangeTelemetry:
+            (value: ConnectionState, oldState: ConnectionState, reason?: string | undefined) => void,
+        private readonly propagateConnectionState: () => void,
+        private readonly isContainerLoaded: () => boolean,
+        private readonly logger: ITelemetryLogger,
+    ) {
+    }
+
+    public receivedAddMemberEvent(clientId: string, quorum: Quorum) {
+        // This is the only one that requires the pending client ID
+        if (clientId === this.pendingClientId) {
+            this.setConnectionState(ConnectionState.Connected);
+        }
+    }
+
+    public receivedDisconnectEvent(reason: string) {
+        this.setConnectionState(ConnectionState.Disconnected, reason);
+    }
+
+    public receivedConnectEvent(
```

Comment:
onSocketConnection ? Or something similar, that is not mentioning delta manager nomenclature?

Reply:
We can use that. Actually, I named these based on the events received. Didn't want to introduce delta manager nomenclature here as this does not deal with delta manager. Like what is done when we receive disconnect event or connect event.